integration/http_test.go

```diff
@@ -0,0 +1,273 @@
+// +build integration
+
+package integration
+
+import (
+	"net/http"
+	"testing"
+
+	"github.com/bitly/go-simplejson"
+	"github.com/infobloxopen/atlas-contacts-app/pkg/pb"
+)
+
+// TestCreateContact_REST uses the REST gateway to create a new contact and
+// ensure the JSON response matches what is expected
+// 1. Create a contact entry with a POST request
+// 2. Unmarshal the JSON into a simplejson struct
+// 3. Ensure the JSON fields match what is expected
+func TestCreateContact_REST(t *testing.T) {
+	dbTest.Reset(t)
+	contact := pb.Contact{
+		FirstName:    "Steven",
+		MiddleName:   "James",
+		LastName:     "McKubernetes",
+		PrimaryEmail: "test@test.com",
+		Notes:        "set sail at sunrise",
+	}
+	res, err := MakeRequestWithDefaults(
+		http.MethodPost,
+		"http://localhost:8080/v1/contacts",
+		contact,
+	)
+	if err != nil {
+		t.Fatalf("unable to create contact: %v", err)
+	}
+	ValidateResponseCode(t, res, http.StatusOK)
+	responseJSON, err := simplejson.NewFromReader(res.Body)
+	if err != nil {
+		t.Fatalf("unable to marshal json response: %v", err)
+	}
+	var tests = []struct {
+		name   string
+		json   *simplejson.Json
+		expect string
+	}{
+		{
+			name:   "contact first name",
+			json:   responseJSON.GetPath("result", "first_name"),
+			expect: `"Steven"`,
+		},
+		{
+			name:   "contact middle name",
+			json:   responseJSON.GetPath("result", "middle_name"),
+			expect: `"James"`,
+		},
+		{
+			name:   "contact last",
+			json:   responseJSON.GetPath("result", "last_name"),
+			expect: `"McKubernetes"`,
+		},
+		{
+			name:   "contact notes",
+			json:   responseJSON.GetPath("result", "notes"),
+			expect: `"set sail at sunrise"`,
+		},
+		{
+			name:   "contact primary email",
+			json:   responseJSON.GetPath("result", "primary_email"),
+			expect: `"test@test.com"`,
+		},
+		{
+			name:   "contact email list",
+			json:   responseJSON.GetPath("result", "emails"),
+			expect: `[{"address":"test@test.com","id":"1"}]`,
+		},
+		{
+			name:   "success response",
+			json:   responseJSON.GetPath("success"),
+			expect: `{"code":"OK","status":200}`,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			ValidateJSONSchema(t, test.json, test.expect)
+		})
+	}
+}
+
+// TestReadContact_REST uses the REST gateway to create a new contact and
+// then read that contact from the application
+// 1. Create a contact entry with a POST request
+// 2. Get the contact from the applicaiton
+// 2. Unmarshal the JSON into a simplejson struct
+// 3. Ensure the JSON fields match what is expected
+func TestReadContact_REST(t *testing.T) {
+	dbTest.Reset(t)
+	contact := pb.Contact{
+		FirstName:    "Wilfred",
+		MiddleName:   "Wallace",
+		LastName:     "O'Docker",
+		PrimaryEmail: "test@test.com",
+		Notes:        "build the containers at dusk",
+	}
+	_, err := MakeRequestWithDefaults(
```

Comment:
you could additionally verify that the response contains the proper fields

Reply:
I verify the behavior of `POST` to `/v1/contacts` in the `TestCreateContact_REST` test, so IMO verifying the response here would be redundant. I'm trying to verify the behavior of `GET` at `/v1/contacts/{id}` with this test.

Reply:
we can agree to disagree on that -- I'm ok if you don't verify it, but it is a different code path and it'd be better to verify it than not. And I think you could do so pretty easily within this code.

Your call :)

Reply:
> it is a different code path

I'm not sure I understand why it's a different code path. Sure, I've changed fields in the contact entry like the name and email, but I haven't changed _how_ the contact gets created (i.e. I haven't exercised any new endpoints, nor have I exercised the same endpoint differently). It seems like added code and complexity that doesn't yield any additional test coverage.

> And I think you could do so pretty easily within this code.

Verifying the fields isn't difficult, but I'd like to avoid writing enormous integration tests.

> Your call :)

That's fair. I'm inclined to move forward with the current approach, but this is _the_ canonical example, so I want to make sure it's halfway decent.

@Evgeniy-L What are your thoughts?

Reply:
I meant it's a different code path from the straight gRPC call in that it gets forwarded through this gateway thing that we're then relying on being correct. There are plenty of potential failure points between the gRPC response that's been tested and this REST response that isn't being tested.

If you don't like it, then I'm gonna submit a change that breaks `POST` responses while keeping the gRPC response intact ;)

By being easy to do I meant that you've already got all your assertions in place for the following GET response and the POST and GET responses should be identical. So you could just iterate over both objects and run the same assertions over them.

Unless I'm wrong about the POST and GET responses being identical, in which case it's either a bug or you should just ignore me and go ahead and merge this.

That being said, I'm still happy with it as-is, so since you've addressed my other feedback already You've got my go-ahead to merge this bad boy.